File: 85-maximal-rectangle/maximal-rectangle.java
class Solution {
    public int largestHistogramRectangle(int[] heights){
        int n = heights.length;
        Stack<Integer>stack = new Stack<>();
        int maxArea=Integer.MIN_VALUE;
        for(int i=0;i<=n;i++){
            while(!stack.isEmpty() && (i==n || heights[i]<=heights[stack.peek()])){
                int height = heights[stack.pop()];
                int width;
                if(stack.isEmpty()){
                    width=i;
                }
                else{
                    width=i-stack.peek()-1;
                }
                maxArea = Math.max(maxArea,width*height);
            }
            stack.push(i);
        }
        return maxArea;
    }
    public int maximalRectangle(char[][] matrix) {
        int n = matrix[0].length;
        int m = matrix.length;
        int height[] = new int[n];
        int maxArea = Integer.MIN_VALUE;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(matrix[i][j]=='0'){
                    height[j]=0;
                    continue;
                }
                else{
                    height[j]++;
                }
            }
            int area = largestHistogramRectangle(height);
            maxArea = Math.max(maxArea,area);
        }
        return maxArea;
    }
}

File: 85-maximal-rectangle/maximal-rectangle.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();
        int maxiarea = 0;
        vector<vector<int>> presum(n, vector<int>(m, 0));

        for (int j = 0; j < m; j++) {
            int sum = 0;
            for (int i = 0; i < n; i++) {
                if (matrix[i][j] == '1') {
                    sum += 1;
                } else {
                    sum = 0;
                }
                presum[i][j] = sum;
            }
        }

        for (int i = 0; i < n; i++) {
            maxiarea = max(maxiarea, lhist(presum[i]));
        }

        return maxiarea;
    }

    int lhist(vector<int>& heights) {
        stack<int> st;
        int max_area = 0;
        int n = heights.size();
        
        for (int i = 0; i <= n; i++) {
            int h = (i == n) ? 0 : heights[i];
            while (!st.empty() && h < heights[st.top()]) {
                int height = heights[st.top()];
                st.pop();
                int width = st.empty() ? i : i - st.top() - 1;
                max_area = max(max_area, height * width);
            }
            st.push(i);
        }
        return max_area;
    }
};